From eecffaecaa6af827b07967d73d6ac52bc612ff9c Mon Sep 17 00:00:00 2001
From: Ramanathan Choodamani <quic_rchoodam@quicinc.com>
Date: Wed, 25 Oct 2023 16:10:08 -0700
Subject: [PATCH] wifi: ath12k: PPEDS- Fix affinity mismatches for ppeds interrupts
on internal pci

Add ppeds_node_idx on the ppeds interrupts to avoid smp affinity
mismatch between network module's edma interrupts and ath12k ppe
interrupts.

Signed-off-by: Ramanathan Choodamani <quic_rchoodam@quicinc.com>
---
 drivers/net/wireless/ath/ath12k/pcic.c | 15 ++++++++-------
 1 file changed, 8 insertions(+), 7 deletions(-)

--- a/drivers/net/wireless/ath/ath12k/pcic.c
+++ b/drivers/net/wireless/ath/ath12k/pcic.c
@@ -591,10 +591,9 @@ int ath12k_pcic_ppeds_register_interrupt
 	u8 bus_id = ab->userpd_id;
 	struct platform_device *pdev = ab->pdev;
 
-	if (type != HAL_REO2PPE && type != HAL_PPE2TCL &&
-	    !(type == HAL_WBM2SW_RELEASE &&
-	    ring_num == HAL_WBM2SW_PPEDS_TX_CMPLN_RING_NUM)) {
-		return 0;
+	if (ab->ppeds_node_idx == -1) {
+		ath12k_err(ab, "invalid ppeds_node_idx in ppeds_register_interrupts\n");
+		return -EINVAL;
 	}
 
 	if (type == HAL_PPE2TCL) {
@@ -603,7 +602,7 @@ int ath12k_pcic_ppeds_register_interrupt
 			goto irq_fail;
 		irq_set_status_flags(irq, IRQ_DISABLE_UNLAZY);
 		snprintf(&ab->dp.ppeds_irq_name[PPEDS_IRQ_PPE2TCL], sizeof(ab->dp.ppeds_irq_name),
-			 "pci%d_ppe2tcl", bus_id);
+			 "pci%d_ppe2tcl_%d", bus_id, ab->ppeds_node_idx);
 		ret = devm_request_irq(&pdev->dev, irq,  ath12k_ds_ppe2tcl_irq_handler,
 				  IRQF_SHARED,
 			    ab->dp.ppeds_irq_name[PPEDS_IRQ_PPE2TCL], (void *)ath12k_dp_get_ppe_ds_ctxt(ab));
@@ -616,7 +615,7 @@ int ath12k_pcic_ppeds_register_interrupt
 			goto irq_fail;
 		irq_set_status_flags(irq, IRQ_DISABLE_UNLAZY);
 		snprintf(&ab->dp.ppeds_irq_name[PPEDS_IRQ_REO2PPE], sizeof(ab->dp.ppeds_irq_name),
-			 "pci%d_reo2ppe", bus_id);
+			 "pci%d_reo2ppe_%d", bus_id, ab->ppeds_node_idx);
 		ret = devm_request_irq(&pdev->dev, irq,  ath12k_ds_reo2ppe_irq_handler,
 				  IRQF_SHARED,
 				  ab->dp.ppeds_irq_name[PPEDS_IRQ_REO2PPE], (void *)ath12k_dp_get_ppe_ds_ctxt(ab));
@@ -629,13 +628,15 @@ int ath12k_pcic_ppeds_register_interrupt
 			goto irq_fail;
 		irq_set_status_flags(irq, IRQ_DISABLE_UNLAZY);
 		snprintf(&ab->dp.ppeds_irq_name[PPEDS_IRQ_PPE_WBM2SW_REL], sizeof(ab->dp.ppeds_irq_name),
-			 "pci%d_ppe_wbm_rel", bus_id);
+			 "pci%d_ppe_wbm_rel_%d", bus_id, ab->ppeds_node_idx);
 		ret = devm_request_irq(&pdev->dev, irq,  ath12k_dp_ppeds_handle_tx_comp,
 				  IRQF_SHARED,
 				  ab->dp.ppeds_irq_name[PPEDS_IRQ_PPE_WBM2SW_REL],(void *)ab);
 		if (ret)
 			goto irq_fail;
 		ab->dp.ppeds_irq[PPEDS_IRQ_PPE_WBM2SW_REL] = irq;
+	} else {
+		return 0;
 	}
 	disable_irq_nosync(irq);
 
